### ISSUE TITLE
Git course (Sully / Cat )

### ISSUE THREAD

**Issue by wise-king-sullyman** [TheOdinProject/top-meta#39]
# Git Course

## Summary

Revise the Git / GitHub related content to increase competency with these tools and equalize the understanding of them between the JS and Ruby tracks.

## Motivation

A lack of Git / GitHub understanding by those on the JS track exemplified the discrepancy between the JS and Ruby tracks. Additionally, students have shown a lack of understanding of the importance of good commit messages, and a lack of knowledge when using Git in a team environment. 

## Suggested Implementation

1: Git Commits
This lesson will explain the importance of commit messages, the best practices for commit messages, when to commit, and how to set VSCode as the default Git commit editor. We are currently planning to put this lesson between the dev tools 1 and Google homepage lessons in Foundations. Once the HTML/CSS rewrite is done it should be easy to port this lesson to before the first project.

2: Basics of Git Branching
This will likely not be substantial enough to be its own lesson, but just a small addition at the top of the RPS project with instructions on how to branch, replacing what was done in [this PR](https://github.com/TheOdinProject/curriculum/pull/22320) which links to [this resource](https://blog.scottlowe.org/2015/01/27/using-fork-branch-git-workflow/) and explains things that are not important to students at the RPS level.

3: PR Procedure & Working with Open Source Projects
This lesson will be largely (if not 100%) based on the [Using Git in the Real World lesson](https://www.theodinproject.com/paths/full-stack-ruby-on-rails/courses/ruby-programming/lessons/using-git-in-the-real-world), but placed somewhere around the Weather App in the JS course and Custom Enumerable in the Ruby course rather than having it in the existing Git section.

4: Advanced Git Topics
This lesson will be largely (if not 100%) based on the [A Deeper Look at Git lesson](https://www.theodinproject.com/paths/full-stack-ruby-on-rails/courses/ruby-programming/lessons/a-deeper-look-at-git), but placed somewhere around Battleship in the JS course and Connect Four in the Ruby course rather than having it in the existing Git section.

## Drawbacks
Adding length to the courses (especially foundations) is always something that should be weighed against the utility added. (Note: If expanding foundations for this content is deemed *not* worthwhile, we believe that a lesson focused on commit messages is still needed and suggest this lesson be after the first projects of the JS and Ruby courses.)

**Comment by kashura** [comment]
I think this is a very solid plan. Improving Git understanding will also give students a bit more dopamine on their github profile. 

One thought that I have: would it make sense to use labs from a tutorial like Git Immersion as part of the assignment? Not all labs, but lets say first 9, or even hand picked specific labs on specific topics?

**Comment by I3uckwheat** [comment]
I think we should mention in this plan that we intend to remove the lessons from Ruby and handle them independently.

**Comment by wise-king-sullyman** [comment]
@kashura I and @CatQueenCodes will have to check out that resource and discuss!

@I3uckwheat edited to try and clarify that the lessons in the existing Git section of the Ruby course will be moved/removed.

**Comment by I3uckwheat** [comment]
Sweet, I like it

**Comment by codyloyd** [comment]
this looks solid.  I like especially that we aren't trying to make some big course to plop in somewhere.. but instead are planning on making some lessons that we can ease in wherever we need them 👌🏽

**Comment by flosker** [comment]
I am planning to start learning with TOP. I see you mentioned "Advanced Git topics" are in [Ruby](https://www.theodinproject.com/paths/full-stack-ruby-on-rails/courses/ruby-programming/lessons/a-deeper-look-at-git) course.

What about students taking JS path? Are there any other lessons, which are placed in specific courses, but are common to both?

**Comment by wise-king-sullyman** [comment]
@flosker this issue is for tracking of an initiative to equalize the Git coverage between the two full stack paths.

There are some other lessons that exist only in the Ruby course at the moment, but they will also likely be added to the JS course before too long.

Also if you aren't already I would encourage you to join our Discord server for discussion of questions like these.

**Comment by CatQueenCodes** [comment]
@kashura @wise-king-sullyman 

GIT Immersion resource overview:

I think that this is a really good learning/teaching resource, however, I don't know if it is a good fit for our curriculum. 

Lessons 1 and 2 involve setting up Git and Ruby. And then the rest of the Git Immersion tutorial uses `.rb` files. 

I think that lessons 1-9 cover most of the basic Git stuff that we have in Foundations, but it does require Ruby to be downloaded and used and I don't think it would be good to confuse learners with that all at that point. I think a long time ago we did have Ruby as part of the installs for everyone at the beginning and that has since been removed and only people doing the Ruby path download it. 

I think some of the more advanced lessons could *possibly* be useful, but if we used them I think they would only be relevant for the students in the Ruby path and it would be nice to keep both Ruby/JS paths having similar content. I don't think there is a JS version of this resource.

I think I would probably vote against using it, but we could possibly pull some ideas from their beginner/advanced lessons and kind of re-write their examples/walkthroughs for some parts if we wanted that worked for both paths.

**Comment by kashura** [comment]
I think in general we have a plan for this one, right? I know ya'll are both busy, so how do we make sure it doesn't go stale or runs out of steam?

**Comment by wise-king-sullyman** [comment]
@kashura I should be able to carve out some time this weekend to get planning done and get the at least started on lesson writing.

**Comment by CatQueenCodes** [comment]
@wise-king-sullyman @kashura ill be camping this weekend but I’ll try and knock out the git commit lesson this month!

**Comment by rlmoser99** [comment]
@wise-king-sullyman @CatQueenCodes I am cleaning up my random notes that I've made and came upon one that I wanted to share with you, but I wasn't sure where to put it because I did not see a ticket for this 3rd lesson.

> 3: PR Procedure & Working with Open Source Projects
> This lesson will be largely (if not 100%) based on the [Using Git in the Real World](https://www.theodinproject.com/paths/full-stack-ruby-on-rails/courses/ruby-programming/lessons/using-git-in-the-real-world) lesson, but placed somewhere around the Weather App in the JS course and Custom Enumerable in the Ruby course rather than having it in the existing Git section.

I once wanted to come up with a way for a person to see both sides of a pull request - creating a pull request and merging a pull request. So my idea is to have a 2-part assignment. First, they solve a code wars problem and post their repo as a project submission. Then the second half of the assignment would be to choose another project submission and create a pull request to add their solution file. 

I know that not everyone would end up getting a PR submitted, but it would give people exposure to going through the motions. We could instruct people to remove their project submission after they had a chance to approve a PR to increase the chances of not having 100's of project submissions with only a few liked submissions getting PRs.

They could even re-use their initial `git_test` repo that they made during the [Git Basics](https://www.theodinproject.com/paths/foundations/courses/foundations/lessons/git-basics) lesson. They would just need to make sure that it was public.

**Comment by wise-king-sullyman** [comment]
@rlmoser99 I **really** like this idea. I do worry a bit about how well it would work in practice though.

I worry that submissions done by people who are no longer active would accumulate and become PR graveyards where the submitter has no idea if they "did it right".

I almost wonder if it might be sustainable to PR to a repo that we maintain for this purpose, where they add a link to a repo for us to PR for them to merge. I know that's similar to the old project submission adding PR process that we used to have which was super unsustainable, but with this being only one lesson and fairly deep in the curriculum I think the numbers might be reasonable, at least until we grow way bigger.

I've got other related concerns and some vague concepts of other ideas for approaches, but I don't think it's worth going more in-depth with them at the moment. I do really like this idea of getting people to experience both sides of PRs.
